### ISSUE TITLE
How to perform L1 regularization constraint training on YOLOv11.

### ISSUE THREAD

**Issue by ZLT-a-LL** [ultralytics/ultralytics#22444]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

请问如何在yolo11中加入L1正则化代码，我是在_do_train()里这样写的，单卡训练时可以正确得到稀疏化模型，但是多卡训练时就失效了，就行完全没有执行一样，没有任何稀疏化效果

<img width="792" height="332" alt="Image" src="https://github.com/user-attachments/assets/a4430d96-3b10-4731-a178-96c194b7c440" />

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ZLT-a-LL, thank you for your interest in Ultralytics 🚀! This is an automated response — an Ultralytics engineer will follow up shortly to assist further.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Since you’re modifying YOLO11 training with L1 regularization and seeing different behavior between single-GPU and multi-GPU, please include:
-Exact ultralytics version and PyTorch/CUDA versions
-The command used to launch training (e.g., `yolo train ...`, `torchrun`, `accelerate`, DDP settings, number of devices)
-The precise location and code snippet of your L1 loss integration (e.g., changes inside `_do_train()`), including where gradients are applied
-Any flags like `--sync-bn`, AMP settings, and seeds
-Relevant logs from both single-GPU and multi-GPU runs showing loss components
-A small MRE repo or script that reproduces the issue on a toy dataset

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by ZLT-a-LL** [comment]
请您给我一个解决方案，真的是困扰良久，使用单卡就可以，多卡就不行

**Comment by Y-T-G** [comment]
You need to install your local repo clone with `pip install -e .`

**Comment by ZLT-a-LL** [comment]
> 您需要使用以下命令安装本地 repo 克隆`pip install -e .`

不是这个问题哦，貌似是多卡情况下梯度同步的问题导致正则化代码根本没有被执行

**Comment by glenn-jocher** [comment]
看起来不是“安装方式”问题，而是 DDP 下正则项只在部分 rank 执行被 allreduce 平均后被稀释了。请确保 L1 在每个 rank 上都参与同一条计算图，并且在主损失回传前叠加，不要放在只在 rank0 执行的日志/保存分支里，例如：

```python
# inside your training step, right before loss.backward()
import torch
import torch.distributed as dist
import torch.nn as nn

lambda_l1 = self.args.get("sparsity", 0.0)
if lambda_l1 > 0:
    l1 = torch.zeros(1, device=loss.device)
    for m in self.model.modules():
        if isinstance(m, nn.BatchNorm2d):  # 常用于稀疏化
            l1 += m.weight.abs().sum()
    # 选一：在所有 rank 上计算（推荐，不要除以 world_size）
    loss = loss + lambda_l1 * l1
    # 选二：若只在 rank0 计算，则需要补偿
    # if dist.is_initialized() and dist.get_rank() == 0:
    #     loss = loss + dist.get_world_size() * lambda_l1 * l1
```

另外请确认多卡是按文档方式启动（如 `device=0,1` 或 `device=[0,1]`），并把正则项放在与主损失同一 `no_sync()`/梯度累积分支内，且不要对正则项调用 `detach()`；参考示例见[Multi‑GPU 训练用法](https://docs.ultralytics.com/modes/train/)。若仍无效，请贴出你的 `_do_train()` 相关片段、实际训练命令以及 `yolo checks` 的文本输出，方便我们排查是否存在多重安装或路径冲突；关于多重安装检测可参见[`check_multiple_install` 参考](https://docs.ultralytics.com/reference/utils/checks/)，常见排障要点也可参考[常见问题排障指南](https://docs.ultralytics.com/guides/yolo-common-issues/)。

**Comment by ZLT-a-LL** [comment]
不好意思，我用您的方式测试了一下，多卡情况下依旧没有稀疏化效果，请问可以再给出一个yolo11使用L1正则化进行约束训练多卡训练的代码示例吗，然后我们后续有集成通用剪枝代码的计划吗

**Comment by glenn-jocher** [comment]
感谢再次测试反馈。多卡下建议用“梯度 hook”给 BN 的缩放参数注入 L1 项，并在 allreduce 平均时用 `world_size` 做补偿；同时兼容 `nn.SyncBatchNorm`，并用 `unwrap_model()` 绕过 DDP/compile 包装，只注册一次即可：

```python
# inside trainer setup (after model is moved to device/DDP), run ONCE per process
import torch
import torch.nn as nn
import torch.distributed as dist
from ultralytics.utils.torch_utils import unwrap_model

alpha = getattr(self.args, "l1_bn", 1e-5)  # your L1 系数
ws = dist.get_world_size() if dist.is_initialized() else 1

base = unwrap_model(self.model)
if not hasattr(base, "_l1_hooks_added"):
    for m in base.modules():
        if isinstance(m, (nn.BatchNorm2d, nn.SyncBatchNorm)):
            p = m.weight
            if p is not None and p.requires_grad:
                # Hook adds ∂(alpha * ||p||_1)/∂p = alpha * sign(p); multiply by world_size to offset grad averaging
                p.register_hook(lambda g, p=p: g + ws * alpha * torch.sign(p))
    base._l1_hooks_added = True
```

如果你之前启用了 `sync_bn=True`，请务必包含 `nn.SyncBatchNorm`，否则在单卡有效、多卡无效的现象就会出现；`unwrap_model()` 参考见[torch_utils.unwrap_model 文档](https://docs.ultralytics.com/reference/utils/torch_utils/).

关于“通用剪枝”集成：目前未内置，欢迎在[Ultralytics Discussions](https://github.com/orgs/ultralytics/discussions)提出需求或提交最小可测的 PR（包含文档与测试）；我们无法承诺时间表，但会按优先级评审。

**Comment by ZLT-a-LL** [comment]
好的，谢谢您，问题已解决！

**Comment by glenn-jocher** [comment]
太好了，感谢更新！为方便后来者，麻烦在此简要说明最终方案或贴上关键片段；若无其他问题请关闭该 Issue；通用剪枝需求可在 Discussions 开新帖或提交带测试与文档的最小 PR，我们会评审。

**Comment by sjtuljw520** [comment]
你好， 可以给一下解决后的代码片段吗？L1正则化的代码建议放在哪个代码文件下面？ @ZLT-a-LL
